### ISSUE TITLE
how can i study you program

### ISSUE THREAD

**Issue by yqtaowhu** [bencz/Compiler-study#1]
hello ,my dear friend
      i very like you program,it is very cool ,but when i read you code ,i find is difficult to me ,please tell 
me how to study compile .
     if you answer me ,i will very very happy

**Comment by bencz** [comment]
Hi :)
What exactly is your doubt?

**Comment by yqtaowhu** [comment]
HI:
   i am very excited you can reply,i want to writen a tiny compiler using c++,i am good at c++,but poor at compiler,now i am study the dragon book,but i find it is diffult to me,can you give some resouses when you study it.
   looking foword to you reply.

**Comment by bencz** [comment]
Hi @yqtaowhu!
Write a compiler, it can be very complex depending on the method you decide to use to perform the parse...
For example, LALR, LL and LR is some parse methods... Each parse system has its own characteristics and there are cases where it is more advantageous to use each analysis system, for example, the [Elena programming language](https://github.com/ELENA-LANG/elena-lang) uses a LALR parse method.
But, in the compiler-study project I use a method called Recursive descent parser, first of all I read all text, and process the text, separating each information and create a AST.
After create the AST the system starts the parse process and in the same time is started to generate the output code.

So basically write a compiler is:
- syntactical analysis
- semantic analysis
- code generation

In the syntactical analysis you just read char by char and do the words/tokens separations, look at this simple sample than I created just now, of how a simple lexer works... [lexer.c](https://gist.github.com/bencz/d8fb165a7bbb9fb9a5ff4b02b322c68e)

**Comment by bencz** [comment]
You can look at some my repositories...

https://gist.github.com/bencz/4116247
https://gist.github.com/bencz/3576dfc8a217a34c05a9
https://gist.github.com/bencz/5174756   -- Maybe you will be very interesting in this code...
https://gist.github.com/bencz/5024780

https://github.com/bencz/asmBuilder
https://github.com/bencz/Bacchi
https://github.com/bencz/dncpp
https://github.com/bencz/IntelAssembler
https://github.com/bencz/Zinnia-lang

Some open source compiler...
https://github.com/LADSoft/OrangeC   --  A COMPLETE C/C++ compiler source
https://github.com/ELENA-LANG/elena-lang -- My favorite project and also, my favorite programming language!

**Comment by yqtaowhu** [comment]
@bencz 
dear friend:
wow,thank you very much ,i am very happy you can write  so clearly,is is very helpful to me,
I will study hard for this...................................
                                                                  wish you happy everyday.

**Comment by bencz** [comment]
@yqtaowhu 
Thanks :)

No problem!
Several people helped me too ...
it does not cost me anything to help others :)
Certainly, the most complex part of developing a compiler is the part of code generation...

**Comment by bencz** [comment]
Look at my old project: https://github.com/bencz/perestroika-compiler
